### PR TITLE
Add Royalty standard

### PR DESCRIPTION
**Pull request by superboyiii**

### CODE REVIEW

**Comment by roman-khimov** [comment]
Standards work when they're standards, that is you say "NEP-11" and everyone has the same understanding of what it is. It's impossible to have this if the definition of "NEP-11" changes. Please, do not change NEP-11 or any other standard. We can have as many standards as needed, this method itself can be a separate standard for example. Or we can obsolete NEP-11 in favor of new NEP-X including this method. It'd be much easier to handle for everyone then.

**Comment by erikzhang** [comment]
But it doesn't break anything and is fully compatible with previous standards. In this case I don't think it's necessary to create a new standard.

**Comment by superboyiii** [comment]
> But it doesn't break anything and is fully compatible with previous standards. In this case I don't think it's necessary to create a new standard.

I agree with Roman, we may add more and more standards on NFT related issues, but we should not add them all in NEP-11 since NEP-11 already has completed. We can't iterate other features on it, NEP-11 doesn't have v1.0, 2.0, 3.0... We should make it clear, different standard should apply on different features, they shouldn't be modified when it's completed. It's also better for developers to know which is doing what.

**Comment by EdgeDLT** [comment]
Agree with the above. NEP-11 is the base class. Some NFTs might want to extend that with additional methods for loyalties, but others may have no reason to implement them. It makes sense to keep the standards separate.

The same is true for https://github.com/neo-project/proposals/issues/152 and NEP-17 tokens, also! We shouldn't want to change NEP-17 to add approve/transferFrom/allowance, but adding a new NEP to standardize these methods for NEP-17 tokens that do want to use that functionality makes complete sense.

**Comment by superboyiii** [comment]
@roman-khimov @erikzhang @EdgeDLT Done.

**Comment by vincentgeneste** [comment]
Great idea to try and make this a standard feature :) we've been actively trying to get this done since so many months!

Couple questions / comments :

Based on the naming of methods / args, i guess you're trying to replicate EIP-2981, but in their case its a per NFT royalty settings, while in your proposal, there's no args, so i assume you want to make it a per contract setting? I would definitely think a per NFT is what's required for this proposal (so the tokenId should have to be passed).

Any reason not to add contractHash in `ReceivedRoyalties` event?

Most importantly, in most places, whether NEO or EVM, all fees are expressed in BPS (1000 -> 10%) directly. I think for convenience purposes and to help onboarding other devs, it should be the same here.

**Comment by superboyiii** [comment]
@merl111 @vincentgeneste  Done.

**Comment by vincentgeneste** [comment]
> @merl111 @vincentgeneste Done.

Couple more comments from me... :)

- your second `royaltyInfo` implementation has safe : false, should be safe : true
- typo `founction`
- we should incorporate multi recipients. While EIP2981 does not support it, it's actually one of the main drawbacks of using it, being limited to only one recipient. So we should allow an array of recipient/values, as well as emit one event per royalty paid
- i do not think `isGlobalRoyalty` is usefull at all. at the end of the day what we want is to know what the royalty is for one NFT. whether it is a global value or not is irrelevent we have to call the contract. So for simplification, i would get rid of it completely and just use your second implementation with `royaltyInfo / tokenId / salePrice`. It is up to the NEP11 contract to define if its global or not, the standard should not really care about it.

**Comment by superboyiii** [comment]
> your second `royaltyInfo` implementation has safe : false, should be safe : true

In second situation, we have a dynamic `royaltyInfo` based on `salePrice`, creator could add his own logic to define how much `royaltyAmount` could be in different `salePrice`, so it should be not safe.

> typo `founction`

Fixed.

> i do not think isGlobalRoyalty is usefull at all. whether it is a global value or not is irrelevent we have to call the contract.

Yes, we have to call the contract, but not necessary to call it for each tokenId. I mean there should be two scenarios: Simple enough or flexible enough. In simple scenario: creators or artists have no requirements for different royalties in different tokenId in one collection, so I'd like to give a simple solution for marketplace to get a clear flag if the royalty is global. It can greatly reduce unnecessary invocation on nodes. If a contract have global royalty, still have to input `tokenId`,  `salePrice` but return the same storage, it's too strange. These params are unnecessary.
For the second scenario, we give a more flexible solution which EIP-2981 already done. (Specific royalty for different tokenId or even dynamic for different salePrice).

**Comment by vincentgeneste** [comment]
> > your second `royaltyInfo` implementation has safe : false, should be safe : true
> 
> In second situation, we have a dynamic `royaltyInfo` based on `salePrice`, creator could add his own logic to define how much `royaltyAmount` could be in different `salePrice`, so it should be not safe.
> 
> > typo `founction`
> 
> Fixed.
> 
> > i do not think isGlobalRoyalty is usefull at all. whether it is a global value or not is irrelevent we have to call the contract.
> 
> Yes, we have to call the contract, but not necessary to call it for each tokenId. I mean there should be two scenarios: Simple enough or flexible enough. In simple scenario: creators or artists have no requirements for different royalties in different tokenId in one collection, so I'd like to give a simple solution for marketplace to get a clear flag if the royalty is global. It can greatly reduce unnecessary invocation on nodes. If a collection(1 contract, different tokenId) have global royalty, still have to input `tokenId`, `salePrice` but return the same storage, it's too strange. These params are unnecessary. For the second scenario, we give a more flexible solution which EIP-2981 already done. (Specific royalty for different tokenId or even dynamic for different salePrice).

Regarding the safe tag, i still think it should be set to true. Even if someone adds his own logic, it should not require write storage access, as we re just trying to get read only data here,i think it's more "safe" to have it set to safe: true :D 

For the `isGlobalRoyalty` i understand the idea behind it personally, it is to propose two scenarios, i just thought personally that this complexified the standard, for very low optimization, but maybe it is just me, i will let others comment on this!

**Comment by superboyiii** [comment]
> Regarding the safe tag, i still think it should be set to true. Even if someone adds his own logic, it should not require write storage access

Yes, you're right, fixed.

> For the `isGlobalRoyalty` i understand the idea behind it personally, it is to propose two scenarios, i just thought personally that this complexified the standard, for very low optimization, but maybe it is just me, i will let others comment on this!

Yes, maybe we can get more feedback to see if it's necessary.

File: nep-x1.mediawiki
Comment (roman-khimov) [inline nep-x1.mediawiki:72]: Why is it a `JSON` if it's clearly a NeoVM array stack item with two elements inside?

File: nep-x1.mediawiki
Comment (roman-khimov) [inline nep-x1.mediawiki:45]: To me, dropping `isGlobalRoyalty` and making just one `royaltyInfo` would make this simpler, global royalty will just have one line implementation of `royaltyInfo`, OK, but without a need for an additional method.

File: nep-x1.mediawiki
Comment (superboyiii) [reply nep-x1.mediawiki]: Done.

File: nep-x1.mediawiki
Comment (superboyiii) [reply nep-x1.mediawiki]: I removed `isGlobalRoyalty` for better compatiblity with EIP-2981

**Comment by vincentgeneste** [comment]
@superboyiii what about supporting multiple recipients ? Is this something we want ? Personally I would love to have it as I see many different use cases for this. It’s not here on EIP2981 , but it’s highly likely because of limitations with multiple transactions in one transaction on EVM, which is not an issue at all on NeoVM.

**Comment by superboyiii** [comment]
@erikzhang @shargon Could you have a review?

**Review by shargon — COMMENTED** [review]
Any method to transfer the royalty to other account?

**Comment by superboyiii** [comment]
> Any method to transfer the royalty to other account?

Yes, this could exist. But since it's customized person by person, we don't need a standard for `set`, we just do it for `get`.

**Comment by superboyiii** [comment]
@erikzhang It's time to move forward.

**Comment by superboyiii** [comment]
> @dotnet-policy-service agree [company="{your company}"]

@dotnet-policy-service agree [company="{Neo}"]

**Comment by superboyiii** [comment]
@erikzhang @shargon Let's move on.

File: nep-x1.mediawiki
Comment (shargon) [inline nep-x1.mediawiki:109]:
```suggestion
<code>royaltyRecipient</code> is the address of who will be sent the royalty payment, MUST be a 20-byte address.
```

Comment (superboyiii) [reply nep-x1.mediawiki]: Done @shargon

**Review by shargon — APPROVED** [review]

**Comment by superboyiii** [comment]
Need votes for going on or not @roman-khimov @shargon @Jim8y @AnnaShaleva @cschuchardt88 @vncoelho

**Review by Jim8y — APPROVED** [review]

**Comment by vncoelho** [comment]
I agree as well, i will do a final read this week.

**Comment by shargon** [comment]
It must change also the readme, isn't it?

**Comment by superboyiii** [comment]
I'm not sure which NEP NO. should pick. If you guys think it's OK, let's decide the NO. And then I will modify readme as well. @shargon

**Comment by cschuchardt88** [comment]
> I'm not sure which NEP No. should pick. If you guys think it's OK, let's decide the No. And then I will modify readme as well. @shargon

It normally is the `github` issue number `id` of the issue that was created. In, this case its would be `NEP-155`.

**Comment by superboyiii** [comment]
> @superboyiii
> 
> > I'm not sure which NEP No. should pick. If you guys think it's OK, let's decide the No. And then I will modify readme as well. @shargon
> 
> It normally is the `github` issue number `id` of the issue that was created. In, this case its would be `NEP-155`.

NEP is a bit special, current standards are a series of number 1 by 1. But I like your idea, this could avoid conflict. @roman-khimov @shargon @Jim8y @AnnaShaleva @cschuchardt88 @vncoelho Is `NEP-155` OK?

**Comment by roman-khimov** [comment]
It doesn't fit current NEP-1, so I doubt it's an improvement. However, this can be a solution to what @Jim8y desperately wants --- reference to unfinished NEP. Can be like NEPR-155 (NEP request 155).

**Comment by superboyiii** [comment]
I think NEO's proposals currently are not frequently created, so we can still follow NEP-1. If more PRs there and iterated quickly, let's make a short proposal like [this](https://github.com/neo-project/proposals/pull/155#issuecomment-1999234326). Now I need `[accepted]` and `[NEP-Standard]` label, so I could ensure it's passed and quickly name it as NEP-23 (now the latest is NEP-22). @roman-khimov @shargon @Jim8y @AnnaShaleva @cschuchardt88 @vncoelho

File: nep-x1.mediawiki
Comment (vncoelho) [inline nep-x1.mediawiki:136]: Here, how is the event for "Multi royaltyRecipient"?

**Review by vncoelho — COMMENTED** [review]
Looks good and precise, just some minor question on the event.

File: nep-x1.mediawiki
Comment (vncoelho) [reply nep-x1.mediawiki:16]: If you want to use this alternative version of the motivation

File: nep-x1.mediawiki
Comment (vncoelho) [inline nep-x1.mediawiki:12]: This NEP defines a global standard to get royalty payment information for Non-Fungible Tokens (NFTs) in order to enable support for royalty payments across all NFT marketplaces in the NEO Smart Economy.

Comment (vncoelho) [reply nep-x1.mediawiki:12]: minor changes here as well, fell free to apply or not

File: nep-x1.mediawiki
Comment (vincentgeneste) [reply nep-x1.mediawiki:136]: It’s one event emitted per recipient

Comment (vncoelho) [reply nep-x1.mediawiki:136]: So fine, thanks

Comment (vncoelho) [reply nep-x1.mediawiki:136]: Let's add this in the description

File: nep-x1.mediawiki
Comment (vncoelho) [inline nep-x1.mediawiki:129]: RoyaltiesTransferred_i

Comment (vncoelho) [reply nep-x1.mediawiki:129]: or RoyaltiesTransferredi
RoyaltiesTransferred1
RoyaltiesTransferred2

File: nep-x1.mediawiki
Comment (vncoelho) [inline nep-x1.mediawiki:121]: Marketplaces that support this standard MUST emit the event for each recipient, <code>RoyaltiesTransferred</code>, after sending a payment.

File: nep-x1.mediawiki
Comment (superboyiii) [reply nep-x1.mediawiki:129]: It's not good for adding a count on 1 event. It's just like `Transfer`, we can have multi `Transfer` in 1 transaction.

**Comment by superboyiii** [comment]
@vncoelho Thanks for your advice, I made some improvement. Please check again.

File: nep-x1.mediawiki
Comment (vncoelho) [reply nep-x1.mediawiki:129]: sure

**Comment by vncoelho** [comment]
> @vncoelho Thanks for your advice, I made some improvement. Please check again.

Looks good to me, fell free to move forward

**Review by vncoelho — APPROVED** [review]

**Review by vncoelho — APPROVED** [review]

**Review by vincentgeneste — APPROVED** [review]

**Comment by shargon** [comment]
Do we change to accept or merge this as a draft and make a different pr for number assignment?

**Comment by vncoelho** [comment]
> Do we change to accept or merge this as a draft and make a different pr for number assignment?

Any is good to me

**Review by Jim8y — APPROVED** [review]

**Comment by roman-khimov** [comment]
> name it as NEP-23 

It's already in https://github.com/neo-project/proposals/pull/156, let's take 24 for it.

**Comment by superboyiii** [comment]
I will update the Number soon!

File: nep-x1.mediawiki
Comment (AnnaShaleva) [inline nep-x1.mediawiki:161]: s/`has 1 events`/`has 1 event`

Comment (AnnaShaleva) [inline nep-x1.mediawiki:161]: s/`if 5 <code>royaltyRecipient</code>`/`if there are 5 <code>royaltyRecipient</code>`

File: nep-x1.mediawiki
Comment (AnnaShaleva) [inline nep-x1.mediawiki:10]: We need to add Table of Contents before the Abstract to match the rest of NEPs.

File: nep-x1.mediawiki
Comment (AnnaShaleva) [inline nep-x1.mediawiki:12]: Need to add a link to https://github.com/neo-project/proposals/blob/master/nep-11.mediawiki.

File: nep-x1.mediawiki
Comment (AnnaShaleva) [inline nep-x1.mediawiki:45]: Should this method always return the same result given the same input data or it's allowed to change royalty amount or receipts through time? Maybe a note should be added. Ref. https://github.com/neo-project/proposals/blob/90686c5fae075541285f1200a584223504878b6d/nep-11.mediawiki?plain=1#L40

File: nep-x1.mediawiki
Comment (AnnaShaleva) [inline nep-x1.mediawiki:161]: Do we have any implementation example? Without implementation this standard can't be moved to the Final stage:

> Implementations -- The implementations must be completed before any NEP is given status "Final", but it need not be completed before the NEP is accepted. It is better to finish the specification and rationale first and reach consensus on it before writing code.

**Review by AnnaShaleva — COMMENTED** [review]
The standard itself looks useful to me.

File: nep-x1.mediawiki
Comment (vincentgeneste) [reply nep-x1.mediawiki:161]: GhostMarket already supports it yes, both our nep11 and our trading contract implemented it. what’s needed to show the implementation? Just link to repo?

Comment (superboyiii) [reply nep-x1.mediawiki:161]: > Do we have any implementation example? Without implementation this standard can't be moved to the Final stage:
> 
> > Implementations -- The implementations must be completed before any NEP is given status "Final", but it need not be completed before the NEP is accepted. It is better to finish the specification and rationale first and reach consensus on it before writing code.

[MegaOasis.io](https://megaoasis.io/) has also implemented this as well.

File: nep-x1.mediawiki
Comment (superboyiii) [reply nep-x1.mediawiki:45]: `royaltyAmount` maybe dynamic. `royaltyRecipient` can be tranferred to another address.

**Comment by superboyiii** [comment]
Updated @AnnaShaleva

**Review by shargon — DISMISSED** [review]

**Review by Jim8y — APPROVED** [review]

**Comment by Jim8y** [comment]
I think this one is good to go @shargon

File: nep-x1.mediawiki
Comment (shargon) [reply nep-x1.mediawiki]: > GhostMarket already supports it yes, both our nep11 and our trading contract implemented it. what’s needed to show the implementation? Just link to repo?

Yes, we could add the link to our contract

Comment (vincentgeneste) [reply nep-x1.mediawiki]: here's our NEP11 implementation https://github.com/OnBlockIO/n3-tokens-contracts/blob/6b01ac3f46b18629f880dcac4e2f5bb4e7f94a37/contracts/NEP11/GhostMarket.NFT.py#L475

Comment (Jim8y) [reply nep-x1.mediawiki]: @superboyiii

File: README.mediawiki
Comment (AnnaShaleva) [inline README.mediawiki:106]: I don't think it should be here. It's not related to this PR. If #156 is merged, then we just rebase onto master.

File: README.mediawiki
Comment (AnnaShaleva) [inline README.mediawiki:154]: Useless change.

File: nep-24.mediawiki
Comment (AnnaShaleva) [inline nep-24.mediawiki:161]: Then we need to add links to the Royalty standard implementation to the end of this NEP, exactly like in https://github.com/neo-project/proposals/blob/master/nep-11.mediawiki#implementation.

File: nep-24.mediawiki
Comment (AnnaShaleva) [inline nep-24.mediawiki:9]: Table of contents is still not added. We need it exactly like in all NEPs, see https://github.com/neo-project/proposals/blob/master/nep-11.mediawiki#table-of-contents.

Comment (Jim8y) [reply nep-24.mediawiki:9]: @superboyiii can make the final format update

Comment (superboyiii) [reply nep-24.mediawiki:9]: > Table of contents is still not added. We need it exactly like in all NEPs, see https://github.com/neo-project/proposals/blob/master/nep-11.mediawiki#table-of-contents.

It's automatically generated. See: https://github.com/neo-project/proposals/blob/8a8b8c229262d46e553b00f9110eae3ccad6c1a3/nep-24.mediawiki
c9ade7d951b311b9529c0c2b6a3c089

Comment (superboyiii) [reply nep-24.mediawiki:9]: Will add another link to C# demo, please don't merge.

File: nep-24.mediawiki
Comment (superboyiii) [reply nep-24.mediawiki:161]: Will add another link to C# demo, please don't merge.

Comment (shargon) [reply nep-24.mediawiki:161]: When i'ts ready, please change the label

**Review by shargon — DISMISSED** [review]

File: nep-24.mediawiki
Comment (AnnaShaleva) [reply nep-24.mediawiki:9]: Got it, my bad.

**Review by AnnaShaleva — COMMENTED** [review]
LGTM, but if we're going to merge it, then I think we need to start with those NEPs that were approved earlier (#156) to avoid conflicts in README.

File: nep-24.mediawiki
Comment (AnnaShaleva) [reply nep-24.mediawiki:161]: Resolved.

File: nep-x1.mediawiki
Comment (AnnaShaleva) [reply nep-x1.mediawiki]: Resolved.

**Comment by superboyiii** [comment]
I think it's not necessary to add count after `royaltyRecipient` and `royaltyAmount` in Multi royaltyRecipient

**Review by shargon — APPROVED** [review]

**Comment by superboyiii** [comment]
@shargon @AnnaShaleva Good to merge!

**Review by shargon — APPROVED** [review]

**Review by AnnaShaleva — APPROVED** [review]

**Review by Jim8y — APPROVED** [review]

**Pull request merged**
